lessons/misc/webdev-intro/contents.md

```diff
@@ -0,0 +1,562 @@
+- Psuedo-classes preceded by a colon
+  - `div:hover` affects a div if the mouse is hovering over it
+  - `h1:first-child` only selects `h1` if it is the first child too
+  - `h1:first-of-type` selects the first `h1` of its siblings
+  - `p:nth-of-type(2n+1)` selects all odd paragraphs
+- Psuedo-elements preceded by two colons `::`
+  - Often used to add decoration ``[href^=http}::after { content: '⤴'; }
+
+## Declarations (`property: value;`)
+
+These describe the way an element is drawn. Different elements can have their own special properties, but no error will be given if a property doesn't apply. 
+
+Common properties include:
+ - `color` : values can be like `red`, `#f00`, `#fa0000`, `rgba(255, 0, 0, .5)`
+ - `background` : as above
+ - `display` : `block`, `inline`, `block-inline`, `flex`, etc.
+ - `margin-top`: a length `0em`, `0.3rem`, `4px`, `1%`
+ - 'visible'
+
+You'll have to look up a reference for all the possibilities.
+
+## Box model
+
+It's worth mentioning the box model for element sizing/spacing. 
+
+<figure>
+![box model illustration](./img/box-model-standard-small.png)
+<figcaption>source: [MDN](https://developer.mozilla.org/en-US/docs/Learn/CSS/Introduction_to_CSS/Box_model)</figcaption>
+</figure>
+
+- `margin`'s are the spaces between objects
+- `border`'s are the size of the drawn border
+- `padding` is the space between the border and the content
+- `width` and `height` describe the size of the element
+  - `box-sizing: content-box` is default where width/height does not include margins, borders or padding.
+  - `box-sizing: border-box` makes width/height include padding and borders (easier math in many cases)
+
+Sites with more complicated layout will often use 
+
+```* { box-sizing: content-box; }```, 
+
+to make the sizes consistent regardless of borders and padding.
+
+## CSS pre-languages
+
+CSS lacks some features that can be very useful/convenient, like
+ - constants
+ - nested rules for hierarchies
+ - conditional logic
+ - include statements
+
+This is good for browser performance, but a pain for developers. There are lots of pre-processors for CSS that add all these features and more
+ - [Stylus](http://stylus-lang.com/) ← *my preference*
+ - [SASS](http://sass-lang.com/) 
+ - [less](http://lesscss.org/)
+
+
+# JavaScript (ECMAScript)
+
+JavaScript is the scripting language supported by modern browsers. It's specified as ECMAScript, the latest standard being ECMAScript 2015 (aka ES6). All modern browsers support ES5.1, and most of [ES6](http://caniuse.com/#search=ES6)
+
+JavaScript is a scripting language with 
+ - C-like syntax (curly braces), 
+ - function scope and [lexical scoping](http://pierrespring.com/2010/05/11/function-scope-and-lexical-scoping/),
+ - dynamic typing,
+ - objects, and
+ - first-class functions
+
+ES6 adds
+ - syntax for classes
+ - template strings with the back tick \` (think  `"""`  in Python, but with `${expressions}`)
```

Comment:
I don't know ES6, but how is `"""` a template string? The f-strings from Python 3.6 might be a better comparison.